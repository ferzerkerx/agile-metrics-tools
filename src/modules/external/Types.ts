export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  srcPath: string;
}

export class ExternalData {
  createdAt: Date;
  [type: string]: any;
}

export interface ExternalService {
  fetchExternalInfo(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
