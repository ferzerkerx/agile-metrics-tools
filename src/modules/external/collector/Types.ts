import { CollectorConfig, MetricItem } from "../../../metrics";

export interface ExternalMetricItem extends MetricItem {}

export class ExternalCollectorConfig implements CollectorConfig {
  since: Date;
  type: string;
  srcType: string;
  srcPath: string;
  metricType: string;
  constructor({ since, type, srcType, srcPath, metricType }) {
    this.since = since;
    this.type = type;
    this.srcType = srcType;
    this.srcPath = srcPath;
    this.metricType = metricType;
  }
}
