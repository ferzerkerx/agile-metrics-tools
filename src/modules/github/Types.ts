export class GithubCommit {
  sha: string;
  createdAt: Date;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  message: string;
  pullRequest?: PullRequest;
}

export class PullRequest {
  id: number;
  numberOfDaysOpen: number;
  numberOfComments: number;
}

export interface GithubConfig {
  repositoryName: string;
  orgName: string;
  since: string;
}

export interface GithubService {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubClient {
  commits(githubConfig: GithubConfig): Promise<string[]>;

  getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<any>;

  pullRequestForCommit(sha: string): Promise<any>;
}
