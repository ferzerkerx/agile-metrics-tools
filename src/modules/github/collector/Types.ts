import { CollectorConfig, MetricItem } from '../../../metrics';

export interface GithubMetricItem extends MetricItem {
  sha: string;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  repositoryName: string;
  pullRequest?: {
    prId: number;
    numberOfDaysOpen: number;
    numberOfComments: number;
  };
}

export class GithubCollectorConfig implements CollectorConfig {
  repositoryName: string;
  orgName: string;
  since: string;

  constructor({ repositoryName, orgName, since }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
  }
}
