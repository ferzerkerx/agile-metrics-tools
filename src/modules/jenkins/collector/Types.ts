import { CollectorConfig, MetricItem } from "../../../metrics/Types";

export interface JenkinsMetricItem extends MetricItem {
  jenkinsUrl: string;
  buildName: string;
  result: string;
  durationInSeconds: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  projectName: string;
}

export class JenkinsCollectorConfig implements CollectorConfig {
  orgName: string;
  projectName: string;
  since: Date;

  constructor({ orgName, projectName, since }) {
    this.orgName = orgName;
    this.projectName = projectName;
    this.since = new Date(since);
  }
}
