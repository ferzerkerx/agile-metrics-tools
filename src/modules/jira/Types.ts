export class Sprint {
  id: number;
  name: string;
  isoStartDate: Date;
  isoEndDate: Date;
}

export class SprintTask {
  key: string;
  typeName: string;
  statusName: string;
  assignee: string;
  epic: string;
  sprint: Sprint;
  created: Date;
  projectName: string;
  teamName: string;
  createdBy: string;
  resolutionDate: Date;
  storyPoints: number;
  labels: string[];
  subtasks: SprintSubtask[];
  histories: { [type: string]: HistoryEntry[] };
  numberOfComments: number;
  numberOfBugs: number;
}

export class SprintSubtask {
  key: string;
  statusName: string;
}

export class HistoryItem {
  field: string;
  fromString: string;
  toString: string;
}

export class HistoryEntry {
  category: string;
  created: Date;
  items: HistoryItem[];
}

export interface JiraConfig {
  teamId: number;
  since: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
}

export interface IssueDetails {
  subtasks: SprintSubtask[];
  histories: { [type: string]: HistoryEntry[] };
  labels: string[];
  created: Date;
  createdBy: string;
  resolutionDate: Date;
  projectName: string;
  teamName: string;
  numberOfComments: number;
  numberOfBugs: number;
}

export interface JiraService {
  completedSprintsSince(teamId: number, referenceDate: Date): Promise<Sprint[]>;

  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<SprintTask[]>;
}

export interface JiraClient {
  getData(url: string): Promise<any>;
}

export interface JiraRepository {
  issueDetailsWithChangelog(
    jiraConfig: JiraConfig,
    issueId: string
  ): Promise<IssueDetails>;
  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<SprintTask[]>;

  completedSprintsSince(teamId: number, referenceDate: Date): Promise<Sprint[]>;
}
