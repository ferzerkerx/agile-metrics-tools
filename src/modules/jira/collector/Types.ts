import { CollectorConfig, MetricItem } from '../../../metrics';

export interface JiraMetricItem extends MetricItem {
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  devTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class JiraCollectorConfig implements CollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban() {
    return this.workFlowType === 'kanban';
  }

  isSprint() {
    return this.workFlowType === 'sprint';
  }
}
