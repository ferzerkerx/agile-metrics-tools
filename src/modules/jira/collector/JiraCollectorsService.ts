import { JiraConfig, JiraService, Sprint, Task } from '../Types';
import { JiraMetricConverter } from './JiraMetricConverter';
import { CollectorService, Utils } from '../../../metrics';
import { JiraCollectorConfig, JiraMetricItem } from './Types';

export class JiraCollectorsService
  implements CollectorService<JiraCollectorConfig, JiraMetricItem> {
  constructor(private readonly jiraService: JiraService) {}

  public async fetch(
    jiraConfig: JiraCollectorConfig
  ): Promise<JiraMetricItem[]> {
    if (!this.supports(jiraConfig)) {
      return [];
    }

    const tasks: Task[] = await this.tasks(jiraConfig);
    return tasks.map((task) =>
      JiraMetricConverter.toMetricItem(jiraConfig, task)
    );
  }

  private async tasks(jiraConfig: JiraCollectorConfig) {
    let tasks = [];
    if (jiraConfig.isKanban()) {
      tasks = await this.tasksForKanban(jiraConfig);
    } else {
      tasks = await this.tasksForSprint(jiraConfig);
    }
    return tasks;
  }

  private async tasksForSprint(
    jiraConfig: JiraCollectorConfig
  ): Promise<Task[]> {
    const sprints: Sprint[] = await this.jiraService.completedSprintsSince(
      jiraConfig.teamId,
      jiraConfig.since,
      jiraConfig.until
    );

    const taskPromises: Promise<Task[]>[] = sprints.map((sprint) =>
      this.sprintDetails(jiraConfig, sprint)
    );

    return Utils.flatMap((item) => item, await Promise.all(taskPromises));
  }

  private async tasksForKanban(
    jiraCollectorConfig: JiraCollectorConfig
  ): Promise<Task[]> {
    const jiraConfig = JiraCollectorsService.toJiraConfig(jiraCollectorConfig);
    return this.jiraService.completedKanbanIssuesSince(jiraConfig);
  }

  private static toJiraConfig(
    jiraCollectorConfig: JiraCollectorConfig
  ): JiraConfig {
    return { ...jiraCollectorConfig };
  }

  private async sprintDetails(
    jiraCollectorConfig: JiraCollectorConfig,
    sprint: Sprint
  ): Promise<Task[]> {
    const jiraConfig = JiraCollectorsService.toJiraConfig(jiraCollectorConfig);
    return this.jiraService.sprintData(jiraConfig, sprint);
  }

  supports(config: any): boolean {
    return config instanceof JiraCollectorConfig;
  }
}
