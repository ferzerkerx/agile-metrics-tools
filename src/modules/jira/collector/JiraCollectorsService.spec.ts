import {
  HistoryEntry,
  JiraConfig,
  JiraService,
  Sprint,
  SprintSubtask,
  SprintTask
} from "../Types";
import { JiraCollectorsService } from "./JiraCollectorsService";
import { JiraCollectorConfig } from "./Types";

function testSubtasks(): SprintSubtask[] {
  return [{ key: "someKey-12", statusName: "Done" }];
}

function testHistories(): { [type: string]: HistoryEntry[] } {
  const historyEntry: HistoryEntry = {
    category: "string",
    created: new Date("2018-12-04"),
    items: [
      {
        field: "status",
        fromString: "Open",
        toString: "Done"
      }
    ]
  };
  return {
    status: [historyEntry]
  };
}

function testSprintTask(): SprintTask {
  return {
    key: "someKey-1",
    typeName: "Story",
    statusName: "Done",
    assignee: "someone",
    epic: "someEpic",
    sprint: testSprint(),
    created: new Date("2018-12-02"),
    projectName: "someProject",
    teamName: "someTeam",
    createdBy: "someone",
    resolutionDate: new Date("2018-12-06"),
    storyPoints: 2,
    labels: ["label1", "label2"],
    subtasks: testSubtasks(),
    histories: testHistories(),
    numberOfComments: 4,
    numberOfBugs: 1
  };
}

function testSprint(): Sprint {
  return {
    id: 45,
    name: "sprint name",
    isoStartDate: new Date("2018-12-03"),
    isoEndDate: new Date("2018-12-11")
  };
}

describe("JiraCollectorsService", () => {
  const jiraService: JiraService = {
    completedSprintsSince: async (
      teamId: number,
      referenceDate: Date
    ): Promise<Sprint[]> => {
      return [testSprint()];
    },

    sprintData: async (
      jiraConfig: JiraConfig,
      sprint: Sprint
    ): Promise<SprintTask[]> => {
      return [testSprintTask()];
    }
  };

  const jiraCollectorsService: JiraCollectorsService = new JiraCollectorsService(
    jiraService
  );

  it("should fetch jiraMetrics", async () => {
    const jiraCollectorConfig: JiraCollectorConfig = new JiraCollectorConfig({
      teamId: 68,
      since: "2018-11-20",
      workFlowMap: null,
      fields: null
    });

    const data = await jiraCollectorsService.fetch(jiraCollectorConfig);
    expect(data).toMatchSnapshot();
  });
});
