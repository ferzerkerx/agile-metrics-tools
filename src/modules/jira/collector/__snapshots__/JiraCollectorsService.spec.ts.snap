// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraCollectorsService should fetch jiraMetrics 1`] = `
Array [
  Object {
    "assignees": Array [
      "someone",
    ],
    "commentCount": 4,
    "createdAt": 2018-12-02T00:00:00.000Z,
    "createdBy": "someone",
    "dataType": "PTS",
    "devTime": 2,
    "estimateHealth": 0,
    "finished": 2018-12-06T00:00:00.000Z,
    "id": "42a10921377014ddde6befd7a8cb74b3e888ac53d78ff051f3ca83abf34a29b9",
    "issueType": "Story",
    "jiraProject": "someProject",
    "key": "someKey-1",
    "leadTime": 4,
    "movedBackward": 0,
    "movedForward": 1,
    "numberOfBugs": 1,
    "rawEstimateHealth": 0,
    "storyPoints": 2,
    "tags": Array [
      "label1",
      "label2",
    ],
    "teamName": "someTeam",
  },
]
`;
